File: 2020/app/src/main/scala/org/snowi/aoc2020/puzzle1/Puzzle1.scala
/* (C) stefan.devops@gmail.com 2021 */
package org.snowi.aoc2020.puzzle1

object Puzzle1 {
  def main(args: Array[String]): Unit = {
    answer()
  }

  def answer(): Unit = {
    val input = readFile("puzzle1/input.txt")
    println()
    println("--------------- Puzzle 01 ---------------")
    println("Result puzzle 2 part 1: " + answer1(input))
    println("Result puzzle 2 part 2: " + answer2(input))
    println()
  }

  def answer1(expenseReport: Array[Int]): Int = {
    var y = 0
    for (x <- expenseReport) {
      for (i <- 0 to (expenseReport.length - 1)) {
        y = expenseReport(i)
        if (x + y == 2020) {
          println(f"$x + $y")
          return x * y
        }
      }
    }
    return -1;
  }

  def answer2(expenseReport: Array[Int]): Int = {
    var y = 0
    var z = 0
    for (x <- expenseReport) {
      for (i <- 0 to (expenseReport.length - 1)) {
        y = expenseReport(i)
        if (x + y < 2020) {
          for (j <- 0 to (expenseReport.length - 1)) {
            z = expenseReport(j)
            if (x + y + z == 2020) {
              println(f"$x + $y + $z = 2020")
              return x * z * y
            }
          }
        }
      }
    }
    return -1;
  }

  def readFile(filename: String): Array[Int] = {
    val bufferedSource = io.Source.fromResource(filename)
    val lines = (for (line <- bufferedSource.getLines()) yield line).toList
    bufferedSource.close
    lines.map(_.toInt).toArray
  }

}

File: 2020/app/src/main/java/org/snowi/aoc2020/puzzle1/Puzzle1.java
/* Copyright: (C) stefan.devops@gmail.com 2021, all rights reserved */
package org.snowi.aoc2020.puzzle1;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.snowi.aoc2020.Puzzle;

import java.util.List;

public class Puzzle1 extends Puzzle {

  private static Logger logger = LoggerFactory.getLogger(Puzzle1.class);

  public Puzzle1(String inputFile) {
    super(inputFile);
  }

  @Override
  public String answer1(List<String> inputFile) {

    int[] expenseReport = toIntArray(inputFile);

    for (int i = 0; i < expenseReport.length; i++) {
      int x = expenseReport[i];
      for (int j = 0; j < expenseReport.length; j++) {
        int y = expenseReport[j];
        logger.debug(" {} + {} = {}", x, y, x + y);
        if (x + y == 2020) {
          return String.valueOf(x * y);
        }
      }
    }
    return "n.a";
  }

  private int[] toIntArray(List<String> inputFile) {
    return inputFile.stream().mapToInt(Integer::parseInt).toArray();
  }

  @Override
  public String answer2(List<String> inputFile) {
    int[] expenseReport = toIntArray(inputFile);

    for (int i = 0; i < expenseReport.length; i++) {
      int x = expenseReport[i];
      for (int j = 0; j < expenseReport.length; j++) {
        int y = expenseReport[j];

        if (x + y < 2020) {
          for (int k = 0; k < expenseReport.length; k++) {
            int z = expenseReport[k];
            logger.debug(" {} + {}  + {} = {}", x, y, z, x + y + z);
            if (x + y + z == 2020) {
              return String.valueOf(x * y * z);
            }
          }
        }
      }
    }
    return "n.a";
  }
}
